From 6c0846b188d83c933f4f497b8d3d65e3a94b8d97 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 31 Mar 2022 20:13:23 +0300
Subject: [PATCH] [SW-77251] gaudi3: add start image api

This patch add start image api for gaudi3 which
will be used by fit operations.

Change-Id: I2d9061d75a6f5b72273984fb4580f8fbc66f7a4c
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 81b7554..7f95521 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -5,6 +5,7 @@
  */
 
 #include <arcboot.h>
+#include <string.h>
 #include <delay.h>
 #include <gaudi3_fw_if.h>
 #include <efuse.h>
@@ -697,3 +698,19 @@
 
 	return false;
 }
+
+void arcboot_start_image(uint64_t startaddr)
+{
+	void (*image_entry)(void);
+	uint32_t arc_addr = (uint32_t)(startaddr);
+
+	if (!startaddr)
+		wfe();
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *)(*(uint32_t *)arc_addr);
+	image_entry();
+}
diff --git a/include/arcboot.h b/include/arcboot.h
index 6ae44e9..09c72d2 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -222,6 +222,7 @@
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 void arcboot_rom_verify(void);
 bool arcboot_is_valid_address_range(uint64_t addr, int len);
+void arcboot_start_image(uint64_t startaddr);
 uint32_t arcboot_error_read_and_clr(uint32_t *major, uint32_t *minor);
 bool arcboot_error_inform(uint32_t major, uint32_t minor);
 bool arcboot_error_is_pending(void);
